ext/CUDAEnzymeCoreExt.jl

```diff
@@ -0,0 +1,91 @@
+module CUDAEnzymeCoreExt
+if isdefined(Base, :get_extension)
+    using EnzymeCore
+    using EnzymeCore.EnzymeRules
+    using CUDA
+else  # using Requires
+    using ..EnzymeCore
+    using ..EnzymeCore.EnzymeRules
+    using ..CUDA
+end
+
+# Expand Enzyme activity and arguments to pass each pointer (primal or shadow) as an indivdual argument
+@inline expand() = ()
+@inline expand(a::Const, args...) = (typeof(a), a.val, expand(args...)...)
+@inline expand(a::Duplicated, args...) = (typeof(a), a.val, a.dval, expand(args...)...)
+@inline expand(a::DuplicatedNoNeed, args...) =
+    (typeof(a), a.val, a.dval, expand(args...)...)
+@inline expand(a::BatchDuplicated, args...) =
+    (typeof(a), a.val, a.dval..., expand(args...)...)
+@inline expand(a::BatchDuplicatedNoNeed, args...) =
+    (typeof(a), a.val, a.dval..., expand(args...)...)
+
+# Contract expanded Enzyme activity and arguments into corresponding structs
+@inline contract() = ()
+@inline contract(::Type{Const{T}}, a, args...) where {T} = (Const(a), contract(args...)...)
+@inline contract(::Type{Duplicated{T}}, a, b, args...) where {T} =
+    (Duplicated(a, b), contract(args...)...)
+@inline contract(::Type{DuplicatedNoNeed{T}}, a, b, args...) where {T} =
+    (DuplicatedNoNeed(a, b), contract(args...)...)
+function metaf(fn, args...)
+    EnzymeCore.autodiff_deferred(Forward, fn, Const, contract(args...)...)
+    nothing
+end
+
+function EnzymeCore.EnzymeRules.forward(
+    ofn::Const{typeof(CUDA.cufunction)},
+    ::Type{<:Duplicated},
+    f::Const{F},
+    tt::Const{TT};
+    kwargs...,
+) where {F,TT}
+    res = ofn.val(f.val, tt.val; kwargs...)
+    return Duplicated(res, res)
+end
+
+function EnzymeCore.EnzymeRules.forward(
+    ofn::Const{typeof(CUDA.cufunction)},
+    ::Type{BatchDuplicated{T,N}},
+    f::Const{F},
+    tt::Const{TT};
+    kwargs...,
+) where {F,TT,T,N}
+    res = ofn.val(f.val, tt.val; kwargs...)
+    return BatchDuplicated(res, ntuple(Val(N)) do _
+        res
+    end)
+end
+
+function EnzymeCore.EnzymeRules.forward(
+    ofn::EnzymeCore.Annotation{CUDA.HostKernel{F,TT}},
+    ::Type{Const{Nothing}},
+    args...;
+    kwargs...,
+) where {F,TT}
+    T2 = (F,)
+    for (p, arg) in zip(TT.parameters, args)
+        if p <: Const
+            T2 = (T2..., Type{typeof(arg)}, p)
+        elseif p <: Duplicated || p <: DuplicatedNoNeed
+            T2 = (T2..., Type{typeof(arg)}, p, p)
+        else
+            T2 = (
+                T2...,
+                Type{typeof(arg)},
+                p,
+                (
+                    ntuple(Val(EnzymeCore.batch_size(typeof(arg)))) do _
+                        p
+                    end
+                )...,
+            )
+        end
+    end
+    TT2 = Tuple{T2...}
```

Comment:
It's not obvious to me what's happening here. Care to add some comments? Otherwise it's going to be hard to fix when something unavoidably breaks.

Reply:
There was a typo in adding batch support, let's see if the commit I added just now from my phone fixes it